// @qlik/sdk/dist/src/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomErrorMsg = exports.CustomError = exports.ConnectionErrorMsg = exports.ConnectionError = exports.AuthenticationErrorMsg = exports.AuthenticationError = void 0;
/* eslint-disable max-classes-per-file */
/**
 * @remarks
 * AuthenticationError is an error that occured during a request
 */
class AuthenticationError extends Error {
}
exports.AuthenticationError = AuthenticationError;
var AuthenticationErrorMsg;
(function (AuthenticationErrorMsg) {
    AuthenticationErrorMsg["FAILED_AUTH_API_KEY"] = "Failed to authenticate host using api-key";
    AuthenticationErrorMsg["UNAUTHENTICATED"] = "Unauthenticated access";
})(AuthenticationErrorMsg = exports.AuthenticationErrorMsg || (exports.AuthenticationErrorMsg = {}));
/**
 * @remarks
 * ConnectionError is an error that occured during creating a request
 */
class ConnectionError extends Error {
}
exports.ConnectionError = ConnectionError;
var ConnectionErrorMsg;
(function (ConnectionErrorMsg) {
    ConnectionErrorMsg["NO_RESPONSE"] = "Connection Error";
})(ConnectionErrorMsg = exports.ConnectionErrorMsg || (exports.ConnectionErrorMsg = {}));
/**
 * @remarks
 * CustomError is an error that occured during setup
 */
class CustomError extends Error {
}
exports.CustomError = CustomError;
var CustomErrorMsg;
(function (CustomErrorMsg) {
    CustomErrorMsg["EMPTY_HOST"] = "Empty host in config";
    CustomErrorMsg["MISSING_CONFIG_PROPERTY"] = "Missing config property";
    CustomErrorMsg["UNSUPPORTED_AUTH_TYPE"] = "Unsupported authType";
    CustomErrorMsg["UNSUPPORTED_PROPERTY"] = "Unsupported property";
    CustomErrorMsg["NOT_IMPLEMENTED"] = "Not implemented";
})(CustomErrorMsg = exports.CustomErrorMsg || (exports.CustomErrorMsg = {}));
