// @qlik/sdk/dist/src/qlik.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Auth = exports.CustomErrorMsg = exports.AuthenticationErrorMsg = exports.ConnectionErrorMsg = exports.AuthType = void 0;
const auth_1 = __importDefault(require("./auth/auth"));
exports.Auth = auth_1.default;
// export types & enums
var types_1 = require("./types/types");
Object.defineProperty(exports, "AuthType", { enumerable: true, get: function () { return types_1.AuthType; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "ConnectionErrorMsg", { enumerable: true, get: function () { return errors_1.ConnectionErrorMsg; } });
Object.defineProperty(exports, "AuthenticationErrorMsg", { enumerable: true, get: function () { return errors_1.AuthenticationErrorMsg; } });
Object.defineProperty(exports, "CustomErrorMsg", { enumerable: true, get: function () { return errors_1.CustomErrorMsg; } });
/**
 * @remarks
 * Qlik.ts is the SDKs entrypoint
 *
 * @param config - required configuration object
 *
 * @returns an instance of the Auth module.
 * @example
 * ```ts
 * import Qlik from '@qlik/sdk';
 * const config: Config =  {
 *   authType: AuthType.WebIntegration,
 *   host: 'my-tenant.qlikcloud.com',
 *   webIntegrationId: '<webintegrationId>',
 * };
 *
 * const qlik = new Qlik(config);
 *
 * if(!qlik.auth.isAuthenticated()){
 *   qlik.auth.authenticate();
 * }
 * ```
 */
class Qlik {
    /** config object reference for the module instance */
    config;
    /** initiated instance of the Authentication module */
    auth;
    /** rest client  reference for the module instance - @experimental */
    rest;
    constructor(config) {
        this.auth = new auth_1.default(config);
        this.rest = this.auth.rest;
    }
}
exports.default = Qlik;
